src/main/java/com/thebois/models/beings/Pawn.java

```diff
@@ -48,4 +59,63 @@ private void setRandomDestination() {
         setPath(newPath);
     }
 
+    protected void updateClosestStructure() {
+        if (closestStructure.isPresent() && getFinalDestination().isPresent()) {
+            // If the structure is completed or destination is unreachable pick a new one
+            if (closestStructure.get().isCompleted() || closestStructure.get().getPosition().equals(
+                getFinalDestination().get())) {
+                closestStructure = findIncompleteStructure();
+                setClosestStructureAsFinalDestination();
+            }
+        }
+        else {
+            closestStructure = findIncompleteStructure();
+            setClosestStructureAsFinalDestination();
+        }
+    }
+
+    protected Optional<IStructure> findIncompleteStructure() {
+        final Collection<Optional<IStructure>> structures =
+            getFinder().findNearestStructures(this.getPosition(), 10);
+
+        for (final Optional<IStructure> structure : structures) {
+            if (structure.isPresent()) {
+                if (!structure.get().isCompleted()) {
+                    return structure;
+                }
+            }
+            else {
+                return Optional.empty();
+            }
```

Comment:
I think there's a logic error here, should it really be giving up and returning `empty()` if the first structure in the collection isn't present? I believe it should instead `continue` to allow the loop to search through the other structures in the list.